File: geomajas/face/geomajas-face-dojo/geomajas-dojo-client/src/main/resources/geomajas/map/print/LegendComponent.js
dojo.provide("geomajas.map.print.LegendComponent");
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
dojo.declare("LegendComponent", BaseComponent, {

	/**
	 * @class 
	 * A legend component (mirror of server object)
	 * @author Jan De Moerloose
	 *
	 * @constructor
	 */
	constructor : function (/* json object */ json, /* unique string */ id) {
		if(!this.mapId){
			this.mapId = "default";
		}
	},
		
	setMapId : function (mapId){
		this.mapId = mapId;
	},
		
	/**
	 * overridden to ignore children (not taken into account for the moment)
	 */
	accept : function (/*Visitor*/visitor, /*Bbox*/ bbox, recursive) {
		visitor.visit (this);
	},

	/**
	 * overridden to ignore children (not taken into account for the moment)
	 */
	executeTopDown : function (/*function*/callback) {
		if(callback(this)){
			return true;
		}
		return false;
	},
	
	/**
	 * overridden to ignore children (not taken into account for the moment)
	 */
	executeBottomUp : function (/*function*/callback) {
		if(callback(this)){
			return true;
		}
	},

	/**
	 * Copies the layers of this map to this component
	 */
	copyLayers : function (mapModel) {
		this.clearItems();
		this.setMapId(mapModel.id);
		var layerList = mapModel.getVisibleLayerList();
		for (var i=0 ; i < layerList.count; i++) {
			var layer = layerList.item(i);
			var item = null;
			if(layer instanceof VectorLayer){
				var label = layer.getLabel();
				var defs = layer.getStyles();
				for (var j = 0; j < defs.count; j++) {
					var text = "";
					var def = defs.item(j);
					if (defs.count > 1) {
						text = label + "(" + def.getName() + ")";
					} else {
						text = label;
					}
					var item = new LegendItemComponent({},this.getId()+"."+i+"."+j);
					item.init(def, text, layer.getLayerType(), this.font);
					this.children.list.push(item);
				}
			} else if(layer instanceof RasterLayer){
				var label = layer.getLabel();
				var item = new LegendItemComponent({},this.id+"."+i);
				item.init(null,label,layer.getLayerType(), this.font);
				this.children.list.push(item);
			}
		}		
	},
	
	/**
	 * Removes all layer components of this map component
	 */
	clearItems : function () {
		for(var i = this.children.list.length-1; i >= 0; i--){
			if(this.children.list[i] instanceof LegendItemComponent){
				this.children.list.splice(i,1);
			}
		}
	}
	
});

File: plugin/geomajas-plugin-printing/src/main/java/org/geomajas/extension/printing/component/LegendComponent.java
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.geomajas.extension.printing.component;

import java.awt.Font;
import java.util.ArrayList;
import java.util.List;

import javax.xml.bind.Unmarshaller;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import org.geomajas.extension.printing.parser.FontAdapter;
import org.geomajas.configuration.client.ClientRasterLayerInfo;
import org.geomajas.configuration.client.ClientVectorLayerInfo;
import org.geomajas.configuration.FeatureStyleInfo;
import org.geomajas.extension.printing.PdfContext;
import org.geomajas.extension.printing.configuration.PrintTemplate;

/**
 * ???
 *
 * @author check subversion
 */
@XmlRootElement
public class LegendComponent extends BaseComponent {

	/**
	 * Application id
	 */
	private String applicationId;
	
	/**
	 * Map id
	 */
	private String mapId;

	public LegendComponent() {
		this("Legend");
	}

	public LegendComponent(String title) {
		this.title = title;
		setConstraint(new LayoutConstraint(LayoutConstraint.RIGHT, LayoutConstraint.BOTTOM,
				LayoutConstraint.FLOW_Y, 0, 0, 20, 20));
		LabelComponent titleLabel = new LabelComponent();
		titleLabel.getConstraint().setAlignmentX(LayoutConstraint.CENTER);
		titleLabel.getConstraint().setMarginY(5);
		titleLabel.setTextOnly(true);
		titleLabel.setText(getTitle());
		titleLabel.setTag(PrintTemplate.TITLE);
		addComponent(titleLabel);
	}

	/**
	 * Call back visitor.
	 *
	 * @param visitor
	 */
	public void accept(PrintComponentVisitor visitor) {
		visitor.visit(this);
	}
	
	/**
	 * The font for the text
	 */
	private Font font = new Font("Dialog", Font.PLAIN, 10);

	/**
	 * Heading text
	 */
	private String title = "Legend";

	@XmlJavaTypeAdapter(FontAdapter.class)
	public Font getFont() {
		return font;
	}

	public void setFont(Font font) {
		this.font = font;
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	@Override
	public void calculateSize(PdfContext context) {
		super.calculateSize(context);
	}

	@Override
	public void render(PdfContext context) {
		// border
		context.fillRectangle(getSize());
		super.render(context);
		// border
		context.strokeRectangle(getSize());
	}

	protected MapComponent getMap() {
		return (MapComponent) getParent();
	}

	public String getMapId() {
		return mapId;
	}

	public void setMapId(String mapId) {
		this.mapId = mapId;
	}
	
	public String getApplicationId() {
		return applicationId;
	}

	
	public void setApplicationId(String applicationId) {
		this.applicationId = applicationId;
	}

	/**
	 * Resets cyclic references like child -> parent relationship.
	 *
	 * @param u
	 * @param parent
	 */
	public void afterUnmarshal(Unmarshaller u, Object parent) {
		setParent((PrintComponent) parent);
	}

	// /**
	// * A legend manages its own children at rendering time, so they shouldn't
	// be
	// * serialized. This magic callback method does the trick !
	// */
	// public void beforeMarshal(Marshaller m) {
	// getChildren().clear();
	// }

	public void clearItems() {
		List<LegendItemComponent> items = new ArrayList<LegendItemComponent>();
		for (PrintComponent child : children) {
			if (child instanceof LegendItemComponent) {
				items.add((LegendItemComponent) child);
			}
		}
		for (LegendItemComponent item : items) {
			removeComponent(item);
		}
	}

	public void addVectorLayer(ClientVectorLayerInfo info) {
		String label = info.getLabel();
		List<FeatureStyleInfo> defs = info.getNamedStyleInfo().getFeatureStyles();
		for (FeatureStyleInfo styleDefinition : defs) {
			String text = "";
			if (defs.size() > 1) {
				text = label + "(" + styleDefinition.getName() + ")";
			} else {
				text = label;
			}
			LegendItemComponent item = new LegendItemComponent(styleDefinition, text, info.getLayerType(), getFont());
			addComponent(item);
		}
	}

	public void addRasterLayer(ClientRasterLayerInfo info) {
		LegendItemComponent item = new LegendItemComponent(null, info.getLabel(), info.getLayerType(),
				getFont());
		addComponent(item);
	}

}
